torch/csrc/autograd/variable.h

```diff
@@ -155,12 +209,26 @@ inline Variable & Variable::grad() {
   return get()->grad;
 }
 
+inline bool Variable::is_leaf() const {
+  return get()->_grad_fn == nullptr;
+}
+
 inline const std::shared_ptr<Function>& Variable::grad_fn() const {
-  return get()->grad_fn;
-};
-inline std::shared_ptr<Function>& Variable::grad_fn() {
-  return get()->grad_fn;
+  return get()->get_grad_fn();
 };
+inline void Variable::set_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
+  assert(!get()->_grad_fn && "set_history can only be called on new Variables");
```

Comment:
Use `TORCH_ASSERT`. Plus, this doesn't really imply that it's a new Variable (it might be an old Variable that only now starts requiring grad).

Reply:
Also, can you add an assertion that `grad_fn` is null if `flags.requires_grad` is false? Same below

Reply:
We still create grad_fn for Python-implemented functions that don't require grad. (I'm going to change that in a subsequent PR).